Sources/MongoSwift/SDAM.swift

```diff
@@ -444,6 +478,84 @@ extension TopologyDescription {
                 return matches
             }
         }
+
+        // If no matches were found during tag set filtering, return an empty list.
         return []
     }
+
+    /// Returns a `Date` representing the latest `lastWriteDate` configured on a secondary in the topology, or `nil`
+    /// if none is found.
+    private func getMaxLastWriteDate() -> Date? {
+        let secondaryLastWriteDates = self.servers.compactMap {
+            $0.type == .rsSecondary ? $0.lastWriteDate : nil
+        }
+        return secondaryLastWriteDates.max()
+    }
+}
+
+extension ServerDescription {
+    /// Calculates the staleness of this server. If the server is not a secondary, the staleness is 0. Otherwise,
+    /// compare against the primary if one is present, or the maximum last write date seen in the topology if present.
+    /// If staleness cannot be calculated due to an absence of values, `nil` is returned.
+    fileprivate func calculateStalenessSeconds(
+        _ primary: ServerDescription?,
+        _ maxLastWriteDate: Date?,
+        _ heartbeatFrequencyMS: Int
+    ) -> Int? {
+        guard self.type == .rsSecondary else {
+            return 0
+        }
+        guard let lastWriteDate = self.lastWriteDate else {
```

Comment:
in what circumstances would a server have no last write date? might be in the specs somewhere but I could not find it in the max staleness one

Reply:
According to the [spec](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#lastwritedate-and-optime), this field is always included in hello responses from 3.4+ servers, so theoretically it should always be present on servers the driver interacts with. If that's the case it would probably make sense to make this field non-optional -- I know we have a ticket in Rust for making more of the fields in our hello response struct non-optional, has that been discussed for Swift? If not, it's something to consider for 2.0.

Reply:
ah thanks for the link - yeah it sounds like we can make it non optional in the future then, could you file a ticket please?